src/Tribe/Values/Abstract_Value.php

```diff
@@ -0,0 +1,455 @@
+<?php
+
+namespace Tribe\Values;
+
+use Tribe\Traits\Reflection_Tools;
+
+abstract class Abstract_Value implements Value_Interface {
+
+	use ValueCalculation;
+	use Reflection_Tools;
+
+	/**
+	 * Holds the initial value passed to the constructor. This variable does not change.
+	 *
+	 * @var mixed
+	 */
+	private $initial_value;
+
+	/**
+	 * Holds the value normalized value calculated when instantiating an object or setting new values.
+	 *
+	 * @var float
+	 */
+	private $normalized_amount;
+
+	/**
+	 * The integer representation of the amount. By default, this is the float value, rounded to the object precision
+	 * places and multiplied by (10^precision).
+	 *
+	 * @var int
+	 */
+	private $integer = 0;
+
+	/**
+	 * The float representation of the amount. By default, this is the same as $normalized_amount
+	 *
+	 * @var float
+	 */
+	private $float = 0.0;
+
+	/**
+	 * The decimal precision to use in calculations.
+	 *
+	 * @var int
+	 */
+	private $precision = 2;
+
+	public $class_name;
+
+	/**
+	 * Initialize class
+	 *
+	 * @since TBD
+	 *
+	 * @param mixed $amount the value to set initially
+	 */
+	public function __construct( $amount = 0 ) {
+		$this->set_initial_representation( $amount );
+		$this->set_normalized_amount( $amount );
+		$this->update();
+		$this->class_name = trim( get_class( $this ), '\\' );
```

Comment:
I would avoid this usage of class completely, it has too many potential problems, I would much rather take a descriptive approach from the class extending this Abstract.

Like give the value a slug of sorts, so that it can be identified as the `type` of value we are talking about.